### ISSUE TITLE
Various java errors

### ISSUE THREAD

**Issue by danboid** [Stefan311/ZfsSpy#7]
Hi Stefan

Great to see someone is working on an open source ZFS recovery tool! Unfortunately I've not been able to view my non-importable filesystem yet.

I've tried ZFSSpy under Ubuntu 20.04 / OpenJDK Runtime Environment (build 11.0.9+11-Ubuntu-0ubuntu1.20.04). My disk is a single disk, unencrypted ZFS pool using lz4 compression that won't import - OpenIndiana says the label is corrupted, ZoL just gives me an IO error but SMART seems to think the drive is OK. I prob made the pool under Ubuntu 18.04 if it matters.

These are the errors I get if I click 'Browse filesystem':

```
Dataset summary:

Errors:
Loading first DMU object failed.
java.lang.Exception: Loading first DMU object failed.
at httpHandlers.HttpHandlerDatasetSummary.createHtmlBody(HttpHandlerDatasetSummary.java:278)
at httpHandlers.HttpHandlerBase.handle(HttpHandlerBase.java:249)
at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
at jdk.httpserver/sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:82)
at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:80)
at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:692)
at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:664)
at jdk.httpserver/sun.net.httpserver.ServerImpl$DefaultExecutor.execute(ServerImpl.java:159)
at jdk.httpserver/sun.net.httpserver.ServerImpl$Dispatcher.handle(ServerImpl.java:442)
at jdk.httpserver/sun.net.httpserver.ServerImpl$Dispatcher.run(ServerImpl.java:408)
at java.base/java.lang.Thread.run(Thread.java:834)
```

I get these errors when I try to looks at some pointers:

```
Errors:
null
java.lang.IndexOutOfBoundsException
at java.base/java.nio.Buffer.checkIndex(Buffer.java:686)
at java.base/java.nio.HeapByteBuffer.get(HeapByteBuffer.java:169)
at dataHandlers.DataHandlerHexdump.printHexdumpToHtml(DataHandlerHexdump.java:57)
at dataHandlers.DataHandlerBonusBuffer.printBonusBufferToHtml(DataHandlerBonusBuffer.java:105)
at httpHandlers.HttpHandlerBlock.getHtmlFromObject(HttpHandlerBlock.java:116)
at httpHandlers.HttpHandlerBlock.createHtmlBody(HttpHandlerBlock.java:181)
at httpHandlers.HttpHandlerBase.handle(HttpHandlerBase.java:249)
at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
at jdk.httpserver/sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:82)
at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:80)
at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:692)
at jdk.httpserver/com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:77)
at jdk.httpserver/sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:664)
at jdk.httpserver/sun.net.httpserver.ServerImpl$DefaultExecutor.execute(ServerImpl.java:159)
at jdk.httpserver/sun.net.httpserver.ServerImpl$Dispatcher.handle(ServerImpl.java:442)
at jdk.httpserver/sun.net.httpserver.ServerImpl$Dispatcher.run(ServerImpl.java:408)
at java.base/java.lang.Thread.run(Thread.java:834)
```

Maybe I'll have more luck using an older Java?

EDIT

Nope, I get the same errors using jre 8 under Ubuntu 18.04.

**Comment by danboid** [comment]
There is definitely hope for us recovering my files off this disk because I can see them using UFS Explorer RAID Recovery but I'd much rather help you get this working than pay £100 for that ;) You can only recover files up to 700k in the demo version of UFS ERR but it lets you browse the filesystem and gives previews for some files.

**Comment by Stefan311** [comment]
Hi, sorry to hear my tool is not working for you. 

I made this tool to recover my own data (Which had finally worked!) . Since I saw that there is no other free tool of this kind, I made it "pretty" and put it on github as a project. I was hoping that other programmers would find the idea interesting and at least share their knowledge of ZFS with me. But no. Silence. The only thing I heard were users who wanted features like RAIDZ support. I'm sorry, but I just don't have the knowledge to do it.

Regarding your case: It is difficult to see from the error messages what exactly is going wrong. The automatic "Browse Filesystem" mode only works when the filesystem is almost functional. For now, let's focus on the manual "Browse Blocks" mode. How far can you get with this mode? The first thing you should see is a list of the 4 copies of the uberblock lists. With the "explore" link you can take a closer look at the Uberblocks-list. If one of them looks completely broken, you can take another. In the Uberblock list you should see a table with transaction numbers at the top. Each transaction represents a file system status. The last - largest - transaction number leads to the most current data, but also to the reason why your filesystem no longer works. It can be useful to search through older transactions. These also contain valid data until they are overwritten bit by bit by other data. If you click on a transaction you will see the individual uberblock with 3 block pointers. If you click on one of these block pointers you will get the first DMU object. In your second error message I can see that it was created when evaluating a DMU object. Is that the case with all DMUs? If so, could you set the "View type" to "Hexdump" and post the hexdump here?
If the DMU view works, you can try to switch to the filesystem mode at this point. Or you can try digging through the different levels of the file system structure. I know the internal file system structure is very complicated. If you want I can try to guide you through these structures.

**Comment by danboid** [comment]
Hi Stefan

Good to hear that you recovered your data using zfsspy! There is certainly an increasing demand for a tool like this and that will only become moreso as Ubuntu improves its ZFS support. ZFS is still mainly a server thing and its kinda expected that you use RAIDZ whereever possible so I'm not surprised you're getting lots of RAIDZ requests, that doesn't mean zfsspy isn't still super useful to people like me.

All copies of the uberblocks list look like this:

![Screenshot at 2020-11-07 10-56-12](https://user-images.githubusercontent.com/1429783/98439399-9093b480-20e9-11eb-86c6-69e56cd9f6e5.png)

One of them is green , the rest are not. A hexdump from on the green transactions pointers is attached.

Thanks for your help - I'll be super impressed if we get this to work!

[hexdump.txt](https://github.com/Stefan311/ZfsSpy/files/5504631/hexdump.txt)

**Comment by danboid** [comment]
Here's a dump of the HTML of one of the pointers
[ZFS Spy.html.tar.gz](https://github.com/Stefan311/ZfsSpy/files/5504637/ZFS.Spy.html.tar.gz)

**Comment by Stefan311** [comment]
Oh that Uberblock table does not look good. As if the table layout have been changed. That explains why you say OpenIndiana says "the label is corrupted". As far I know, OpenIndia uses the outdated original ZFS implementation from SUN. So it may have the same problem as ZfsSpy. I fear ZfsSpy have to be heavily upgraded to match the new layout. 

I am really sorry, but I will not do heavy work on this project any more.
Maybe you can use the official zdb debug tool to get you data back.
Some interesting conversations about this topic can be found here: https://groups.google.com/g/mailing.freebsd.hackers/c/QzQA32BmaqQ/m/AoLEZrfkElMJ

**Comment by danboid** [comment]
No problem Stefan, thanks for releasing what you have. Hopefully someone who knows more about ZFS will fork it and make is usable for more situations. Thankfully, I have been able to recover my files with UFS Explorer but I would've preferred to do it using open source tools if possible. I had already tried those commands you suggested as well as many like them with no luck under Linux, FreeBSD and OpenIndiana.
